### PR TITLE
sTLOS Staking UI

### PR DESCRIPTION
**Pull request by EJSG**
## PR Notes
This PR introduces Staking functionality to Teloscan, allowing the user to interact with the new STLOS and TLOS Escrow contracts. Included in this version:

- view stats about staking on Telos, as well as staking-related info for logged in account
- stake TLOS in exchange for sTLOS
- unstake TLOS to escrow
- claim unlocked un-staked TLOS
- fixed a bug in `FuntionInterface.vue` relating to payable methods
- updated app footer spacing
- added copy button to `TransactionField.vue` (ie. all truncated transaction hashes sitewide)
- added new Sass partial `_utilities` for utility classes, including various flexbox presets
  (eg. `u-flex--center-y`)
- updated sitewide appearance (reduced `border-radius` on buttons and containing divs)
- removed some defunct global styles
- updated Vue template linting rules (max attributes per line)
- added npm script `nuke` (invoked via `yarn nuke`), which deletes node modules + their caches, then re-installs them

## Testing notes
**Setup**
- if you are running this code locally:
  - pull this branch
  - run `yarn nuke`
  - add the following testnet addresses to your `.env` file:
    ```
     STAKED_TLOS_CONTRACT_ADDRESS=0xF200706fe04c3B365A9BcB466e1939B9d524057E
     TELOS_ESCROW_CONTRACT_ADDRESS=0x7E9cF9fBc881652B05BB8F26298fFAB538163b6f
    ```
  - run `yarn install && yarn dev` in the project root
- if you are not running this code locally:
    - replace `http://localhost:8080` in these instructions with https://stage.teloscan.io
    - note: you may not be able to access vue devtools when using this link
- if you do not already have 2+ accounts linked to Teloscan, make new account (needed to test account switching). if you are using MetaMask this can be done by:
    1. clicking metamask logo in your browser
    2. click the profile picture of your current account (top right)
    3. click _Create Account_
- add funds from faucet to main account if needed (make sure to leave an account with a 0 TLOS balance)
    1. copy your account address (can be done by clicking MetaMask => copy button at the top)
    2. go to https://app.telos.net/testnet/developers
    3. paste address in the lowest input w/ placeholder "Send to EVM Address"
    5. click _Send Testnet EVM TLOS_ button
- make sure Vue devtools for vue3 are added to your browser: https://devtools.vuejs.org/
    - once installed, open browser devtools from http://localhost:8080
    - enable 'editable props' in Vue devtools settings

**Responsive Layout**
- go to http://localhost:8080/staking
- resize the browser window from sizes XS (mobile) to your maximum monitor size
    - onscreen elements should reflow and remain legible at all sizes
    - note that some elements may look too small given testnet values, but will grow to fill up space in prod (especially at small and medium breakpoints, which is the tradeoff I made given their lower usage compared to xs and lg+), e.g. the `TVL` stat will quickly grow to millions

**Light/dark theme**
- go to http://localhost:8080/staking
- toggle light/dark mode using the Sun/Moon icon in the navbar
- click around the page
    - all text should be legible

**Tabs/Routing**
- go to http://localhost:8080/staking#claim
    - the page should load with the _Claim_ tab selected
- go to http://localhost:8080/staking#unstake
    - the page should load with the _Unstake_ tab selected
- go to http://localhost:8080/staking#stake
    - the page should load with the _Stake_ tab selected
- go to an invalid hash, eg. http://localhost:8080/staking#they-shoulda-called-it-antelopEOS
    - the page should reroute to http://localhost:8080/staking#stake

**Input masking**
- ensure you are logged out
    - top right of the app should say "Connect Wallet"
- go to http://localhost:8080/staking
- open your browser devtools
    - throughout this process, no input should cause a console error
- switch to _Network_ tab
    - as you type in characters, you should only see one request for each group of characters you enter
    - …which is to say, calls to convert TLOS to/from STLOS should be debounced and not generating a request for each character typed
- enter random / invalid characters to either onscreen text input, eg. special characters, letters, emojis
    - the input should only recognize integers and 0 or 1 decimal points
- type a number followed by as many decimals as possible, eg. 1.23456789.....
    - you should be able to enter up to 18 decimal places after a `.`
- clear the input and type a single `.`
    - a `0` should be inserted before the `.`, ie. `0.`
    - the Stake CTA should be disabled for an input value of `0`
- clear input and type `0`
    - the Stake CTA should be disabled
- clear input and type `0.` followed by any number of zeroes, eg. `0.0000000`
    - the Stake CTA should be disabled
- clear the input
    - the Stake CTA should be disabled
- enter a number with a decimal, eg. `123.456`
    - place your cursor _after_ the decimal and then backspace it
    - the period `.` should be deleted and `123,456` should take its place
- try the same operation again, this time placing the cursor _before_ the decimal and using the `Del` key (or equivalent delete-forward depending on OS)
    - the period `.` should be deleted and `123,456` should take its place
- clear the input and enter a value with commas, eg. `123,456,789.098`
    - place your cursor after a comma, such as directly before the `7` in the example above
    - backspace
    - the number across the comma (`6` in this case) should be deleted, resulting in `1,245,789`
    - repeat, again using delete-forward
- type any arbitrary string into your address bar, eg. `78g9..%DROP_TABLE_USERS&&722f]]`
- copy and paste string w/ illegal characters into the input
    - only the intergers in the string should paste, `789,722` for the above example
- clear the input and type a bunch of numbers in a row, eg. `999999999999999999999999999999999999`
    - when you reach the maximum (which is an arbitrary large number when not signed in; signed in values will be tested later), the input should wiggle and disallow you from adding more digits

**Account handling**
- go to http://localhost:8080/staking
- ensure you are logged out
    - other than APY and TVL, stats at the top of the page should all display`--`
    - the top input should show the text "Wallet not connected"
    - the CTA should show the text "Connect Wallet"
- switch to the _Unstake_ tab
    - CTA and input message should be the same
- switch to the _Claim_ tab
    - CTA should be disabled
    - table should display no data
- switch back to _Stake_ tab
- enter a very large value into one or both of the inputs, eg. `999999999999999999`
- click the CTA ("Connect wallet")
    - login modal should pop open
- sign in as normal
    - stats at the top of the page should refresh and reflect your account balances
    - top input info text should now display your account TLOS balance
    - the inputs should reset (ie. no longer display `9999....9`)
    - if you are using MetaMask, you should see a link at the top of the page to add sTLOS to MetaMask
        - clicking this link should pop a MetaMask window asking to confirm tracking sTLOS
- switch to an account with 0 TLOS balance
    - stats at the top of the page should refresh and display `0.000`
    - top input should display `Insufficient TLOS balance to stake`
    - CTA should say "Get more TLOS"
- click CTA
    - you should be brought, in a new window, to trade for TLOS

**Stake TLOS**
- go to http://localhost:8080/staking
- ensure you are signed into an account with a nonzero account balance
- attempt to type in a value larger than wallet balance, eg. 9999999
    - you should receive visual feedback (input wiggle) and the input value should be set to your wallet TLOS balance _minus 1 TLOS_
- clear the input
- click the value displayed in the top right corner of the input (your usable wallet balance)
    - this value should be placed in the input with full precision (leading `0`s will be trimmed)
- clear the input
- enter a small amount of TLOS (small enough that you can do several of these tests with your current balance)
- click _Stake_ CTA
    - a modal should pop informing you of the terms of staking
    - clicking outside of the modal, or clicking cancel, should close the modal
- click Stake in the modal that pops
- CTA should be disabled and say "Loading..." while MetaMask prompt is open
- confirm the transaction on MetaMask
    - stats and balances should refresh
    - a link to view the transaction should appear below the form
- go to _Unstake_ tab
    - the balance in the top input should match your sTLOS balance

**Unstake sTLOS**
- go to http://localhost:8080/staking#unstake
- ensure you are signed into an account with a nonzero account balance
- attempt to type in a value larger than account sTLOS balance, eg. 9999999
    - you should receive visual feedback (input wiggle) and the input value should be set to your sTLOS balance
- clear the input
- click the value displayed in the top right corner of the input (your sTLOS balance)
    - this value should be placed in the input with full precision (leading `0`s will be trimmed)
- clear the input
- enter a small amount of sTLOS (small enough that you can do several of these tests with your current balance)
- click _Untake_ CTA
    - a modal should pop informing you of the terms of unstaking
    - clicking outside of the modal, or clicking cancel, should close the modal
- click Unstake in the modal that pops
- CTA should be disabled and say "Loading..." while MetaMask prompt is open
- confirm the transaction on MetaMask
    - stats and balances should refresh
    - a link to view the transaction should appear below the form
- go to _Unstake_ tab
    - the balance in the top input should match your sTLOS balance
- go to _Claim_ tab
    - the transaction should show in the table
- go back to the _Unstake_ tab
- open Vue devtools
- search in vue devtools for component `UnstakeForm`
- modify the `data` property `maxDeposits` to be `9`
- enter another TLOS value and click _Unstake_
  - the modal which pops should now include a warning about the maximum number of concurrent unstakes
- cancel the unstake

**View/Claim Escrowed TLOS**
- after unstaking TLOS in the previous section, go stretch your legs. You will need to wait for the unstake period to elapse (10 mins on testnet)
    - note that, because the mainnet unstake period will be closer to 10 days, there is no timer set to notify user that the unstake period has elapsed
- after waiting, go to http://localhost:8080
    - you should see a banner notifying you that you have TLOS to claim
    - there should be a green dot on the 
- click Dismiss on the banner
    - the banner should disappear
    - the green dot should still be visible on the Claim tab
- reload the page
  - the banner should re-appear
- click "Claim" on the banner
  - you should be taken to the Claim tab
- _before claiming your TLOS_, go back and Unstake (Staking more first if needed) some amount TLOS
- go to http://localhost:8080/staking#claim
  - you should see your locked and unlocked TLOS balances along with their claim date
  - your unlocked TLOS should have a date in the past, eg. "5 days ago"
- click Claim
- authenticate the transaction
  - table of withdrawals should update
  - a link to view the claim tx should appear
  - your personal stats at the top of the page should update

### CODE REVIEW

File: src/components/Footer.vue [hunk 1]
```diff
@@ -35,9 +35,9 @@ export default {
           a(
             href="https://eosauthority.com/vote/producers?network=telos"
             target="_blank"
-          ) Yknot &
+          ) &nbsp;Yknot &
```
Comment (EJSG) [inline src/components/Footer.vue:38]: improve spacing on the footer text

**Comment by Kersten-TCD** [comment]
Performed tests:

Claim section is very hard to read in "day-mode" please adjust color scheme to be better readable.

Bildschirmfoto 2022-08-25 um 11 16 36

For multiple claims it is unclear which position you will claim once hitting the button. I also think we discussed that the function "Claim Tlos" would claim all positions possible.

Testing this further I had 3 positions ready for claim, I triggered claim and 2 of 3 positions were removed immediately: TX: 0xcb5ab625cea8ee4c13f985ff620774441536cc03243c995f929969a4e735c21b. One remaining had to be claimed in different transaction. TX: 0x49c39f03da6096212ef9ed82fa18fcaa22d2ca434153b9a2564f6d6de91425da

Bildschirmfoto 2022-08-25 um 11 19 25

**Comment by Kersten-TCD** [comment]
Unexpected behavior, unstake of 1 sTLOS only reduces the available sTLOS balance by 0.1 and not by 1.0 see there consecutive transactions.

[16:05] Thomas Cuvillier
yes because instead of trying to withdraw 11.15 TLOS (the equivalent of 1 sTLOS), the front end tries to withdraw 1 TLOS (the equivalent of 0.1 sTLOS)


Bildschirmfoto 2022-08-29 um 15 49 44
Bildschirmfoto 2022-08-29 um 15 50 03
Bildschirmfoto 2022-08-29 um 15 50 28

Proposal to add comment to staking modal educating the users that the amount of sTLOS will be significantly lower than the amount of TLOS entered into "Stake TLOS" field. This will prevent confusion of users, same behavior applied to specific liquidity pools on Omnidex and Zappy which caused various user requests in the past.

Bildschirmfoto 2022-08-29 um 16 07 21

**Comment by donnyquixotic** [comment]
> Unexpected behavior, unstake of 1 sTLOS only reduces the available sTLOS balance by 0.1 and not by 1.0 see there consecutive transactions.
> 
> [16:05] Thomas Cuvillier yes because instead of trying to withdraw 11.15 TLOS (the equivalent of 1 sTLOS), the front end tries to withdraw 1 TLOS (the equivalent of 0.1 sTLOS)
> 
> Bildschirmfoto 2022-08-29 um 15 49 44 Bildschirmfoto 2022-08-29 um 15 50 03 Bildschirmfoto 2022-08-29 um 15 50 28
> 
> Proposal to add comment to staking modal educating the users that the amount of sTLOS will be significantly lower than the amount of TLOS entered into "Stake TLOS" field. This will prevent confusion of users, same behavior applied to specific liquidity pools on Omnidex and Zappy which caused various user requests in the past.
> 
> Bildschirmfoto 2022-08-29 um 16 07 21

RE conversion explainer, this may be a moot point. we are adding an explicit conversion value (e.g. 1 TLOS = .3 sTLOS) to the UI.

**Comment by donnyquixotic** [comment]
@Kersten-TCD RE conversion explainer, We are adding an explicit conversion value (e.g. 1 TLOS = .3 sTLOS) to the UI so we may not need an explainer. We can still add it if we want but probably unnec if the user can see the conversion right on the form.

**Review by donnyquixotic — DISMISSED** [review]

**Review by donnyquixotic — DISMISSED** [review]

**Comment by donnyquixotic** [comment]
@EJSG  I set up a temporary instance just for this branch that will automatically update:
https://staking-teloscan.netlify.app/

**Review by poplexity — APPROVED** [review]
👍

**Pull request merged**